<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="The HIMA Development Team" />

<meta name="date" content="2025-06-11" />

<title>High-Dimensional Mediation Analysis</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">High-Dimensional Mediation Analysis</h1>
<h3 class="subtitle">A Guide to Using the HIMA Package</h3>
<h4 class="author">The HIMA Development Team</h4>
<h4 class="date">2025-06-11</h4>


<div id="TOC">
<ul>
<li><a href="#introduction" id="toc-introduction">Introduction</a></li>
<li><a href="#package-overview" id="toc-package-overview">Package
Overview</a></li>
<li><a href="#data-preparation-and-settings" id="toc-data-preparation-and-settings">Data Preparation and Settings</a>
<ul>
<li><a href="#hima-function-interface" id="toc-hima-function-interface"><code>hima</code> Function
Interface</a></li>
<li><a href="#formula-argument-formula" id="toc-formula-argument-formula">1. Formula Argument
(<code>formula</code>)</a></li>
<li><a href="#phenotype-data-data.pheno" id="toc-phenotype-data-data.pheno">2. Phenotype Data
(<code>data.pheno</code>)</a></li>
<li><a href="#mediator-data-data.m" id="toc-mediator-data-data.m">3.
Mediator Data (<code>data.M</code>)</a></li>
<li><a href="#about-data-scaling" id="toc-about-data-scaling">4. About
data scaling</a></li>
<li><a href="#parallel-computing-support" id="toc-parallel-computing-support">Parallel Computing Support</a>
<ul>
<li><a href="#enabling-parallel-computing" id="toc-enabling-parallel-computing">Enabling Parallel
Computing</a></li>
</ul></li>
</ul></li>
<li><a href="#applications-and-examples" id="toc-applications-and-examples">Applications and Examples</a>
<ul>
<li><a href="#load-the-hima-package" id="toc-load-the-hima-package">Load
the <code>HIMA</code> Package</a></li>
<li><a href="#continuous-outcome-analysis" id="toc-continuous-outcome-analysis">Continuous Outcome
Analysis</a></li>
<li><a href="#efficient-hima" id="toc-efficient-hima">Efficient
HIMA</a></li>
<li><a href="#binary-outcome-analysis" id="toc-binary-outcome-analysis">Binary Outcome Analysis</a></li>
<li><a href="#survival-outcome-analysis" id="toc-survival-outcome-analysis">Survival Outcome Analysis</a></li>
<li><a href="#microbiome-mediation-analysis" id="toc-microbiome-mediation-analysis">Microbiome Mediation
Analysis</a></li>
<li><a href="#quantile-mediation-analysis" id="toc-quantile-mediation-analysis">Quantile Mediation
Analysis</a></li>
</ul></li>
</ul>
</div>

<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>Mediation analysis is a statistical method used to explore the
mechanisms by which an independent variable influences a dependent
variable through one or more intermediary variables, known as mediators.
This process involves assessing the indirect, direct, and total effects
within a defined statistical framework. The primary goal of mediation
analysis is to test the mediation hypothesis, which posits that the
effect of an independent variable on a dependent variable is partially
or fully mediated by intermediate variables. By examining these
pathways, researchers can gain a deeper understanding of the causal
relationships among variables. Mediation analysis is particularly
valuable in refining interventions to enhance their effectiveness in
clinical trials and in observational studies, where it helps identify
key intervention targets and elucidate the mechanisms underlying complex
diseases.</p>
</div>
<div id="package-overview" class="section level1">
<h1>Package Overview</h1>
<p>The <code>HIMA</code> package provides robust tools for estimating
and testing high-dimensional mediation effects, specifically designed
for modern <code>omic</code> data, including epigenetics,
transcriptomics, and microbiomics. It supports high-dimensional
mediation analysis across continuous, binary, and survival outcomes,
with specialized methods tailored for compositional microbiome data and
quantile mediation analysis.</p>
<p>At the core of the package is the <code>hima</code> function, a
flexible and powerful wrapper that integrates various high-dimensional
mediation analysis methods for both effect estimation and hypothesis
testing. The <code>hima</code> function automatically selects the most
suitable analytical approach based on the outcome and mediator data
type, streamlining complex workflows and reducing user burden.</p>
<p><code>hima</code> is designed with extensibility in mind, allowing
seamless integration of future features and methods. This ensures a
consistent, user-friendly interface while staying adaptable to evolving
analytical needs.</p>
</div>
<div id="data-preparation-and-settings" class="section level1">
<h1>Data Preparation and Settings</h1>
<p>The <code>hima</code> function provides a flexible and user-friendly
interface for performing high-dimensional mediation analysis. It
supports a variety of analysis methods tailored for continuous, binary,
survival, and compositional data. Below is an overview of the
<code>hima</code> function and its key parameters:</p>
<div id="hima-function-interface" class="section level3">
<h3><code>hima</code> Function Interface</h3>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">hima</span>(</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a>  formula,          <span class="co"># The model formula specifying outcome, exposure, and covariate(s)</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a>  data.pheno,       <span class="co"># Data frame with outcome, exposure, and covariate(s)</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a>  data.M,           <span class="co"># Data frame or matrix of high-dimensional mediators</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a>  mediator.type,    <span class="co"># Type of mediators: &quot;gaussian&quot;, &quot;negbin&quot;, or &quot;compositional&quot;</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a>  <span class="at">penalty =</span> <span class="st">&quot;DBlasso&quot;</span>,  <span class="co"># Penalty method: &quot;DBlasso&quot;, &quot;MCP&quot;, &quot;SCAD&quot;, or &quot;lasso&quot;</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a>  <span class="at">quantile =</span> <span class="cn">FALSE</span>, <span class="co"># Use quantile mediation analysis (default: FALSE)</span></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a>  <span class="at">efficient =</span> <span class="cn">FALSE</span>,<span class="co"># Use efficient mediation analysis (default: FALSE)</span></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a>  <span class="at">scale =</span> <span class="cn">TRUE</span>,     <span class="co"># Scale data (default: TRUE)</span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a>  <span class="at">sigcut =</span> <span class="fl">0.05</span>,    <span class="co"># Significance cutoff for mediator selection</span></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a>  <span class="at">contrast =</span> <span class="cn">NULL</span>,  <span class="co"># Named list of contrasts for factor covariate(s)</span></span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a>  <span class="at">subset =</span> <span class="cn">NULL</span>,    <span class="co"># Optional subset of observations</span></span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a>  <span class="at">verbose =</span> <span class="cn">FALSE</span>   <span class="co"># Display progress messages (default: FALSE)</span></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a>)</span></code></pre></div>
<p>To use the <code>hima</code> function, ensure your data is prepared
according to the following guidelines:</p>
</div>
<div id="formula-argument-formula" class="section level3">
<h3>1. Formula Argument (<code>formula</code>)</h3>
<p>Define the model formula to specify the relationship between the
<code>Outcome</code>, <code>Exposure</code>, and
<code>Covariate(s)</code>. Ensure the following:</p>
<ul>
<li><p><strong>General Form</strong>: Use the format
<code>Outcome ~ Exposure + Covariate(s)</code>. Note that the
<code>Exposure</code> variable represents the exposure of interest
(e.g., “Treatment” in the demo examples) and it has to be listed as the
first independent variable in the formula. <code>Covariate(s)</code> are
optional.</p></li>
<li><p><strong>Survival Data</strong>: For survival analysis, use the
format <code>Surv(time, event) ~ Exposure + Covariate(s)</code>. See
data examples <code>SurvivalData$PhenoData</code> for more
details.</p></li>
</ul>
</div>
<div id="phenotype-data-data.pheno" class="section level3">
<h3>2. Phenotype Data (<code>data.pheno</code>)</h3>
<p>The <code>data.pheno</code> object should be a
<code>data.frame</code> or <code>matrix</code> containing the phenotype
information for the analysis (without missing values). Key requirements
include:</p>
<ul>
<li><p><strong>Rows:</strong> Represent samples.</p></li>
<li><p><strong>Columns:</strong> Include variables such as the outcome,
treatment, and optional covariate(s).</p></li>
<li><p><strong>Formula Consistency:</strong> Ensure that all variables
specified in the <code>formula</code> argument (e.g.,
<code>Outcome</code>, <code>Treatment</code>, and
<code>Covariate(s)</code>) are present in
<code>data.pheno</code>.</p></li>
</ul>
</div>
<div id="mediator-data-data.m" class="section level3">
<h3>3. Mediator Data (<code>data.M</code>)</h3>
<p>The <code>data.M</code> object should be a <code>data.frame</code> or
<code>matrix</code> containing high-dimensional mediators (without
missing values). Key requirements include:</p>
<ul>
<li><p><strong>Rows:</strong> Represent samples, aligned with the rows
in <code>data.pheno</code>.</p></li>
<li><p><strong>Columns:</strong> Represent mediators (e.g., CpGs, genes,
or other molecular features).</p></li>
<li><p><strong>Mediator Type:</strong> Specify the type of mediators in
the <code>mediator.type</code> argument. Supported types include:</p>
<ul>
<li><code>&quot;gaussian&quot;</code> for continuous mediators (default, e.g., DNA
methylation data).</li>
<li><code>&quot;negbin&quot;</code> for count data (e.g., transcriptomic
data).</li>
<li><code>&quot;compositional&quot;</code> for microbiome or other compositional
data.</li>
</ul></li>
</ul>
</div>
<div id="about-data-scaling" class="section level3">
<h3>4. About data scaling</h3>
<p>In most real-world data analysis scenarios, <code>scale</code> is
typically set to <code>TRUE</code>, ensuring that the exposure (variable
of interest), mediators, and covariate(s) (if included) are standardized
to a mean of zero and a variance of one. No scaling will be applied to
<code>Outcome</code>. However, if your data is already
pre-standardized—such as in simulation studies or when using our demo
dataset-<code>scale</code> should be set to <code>FALSE</code> to
prevent introducing biases or altering the original data structure.</p>
<p>When applying <code>HIMA</code> to simulated data, if
<code>scale</code> is set to <code>TRUE</code>, it is imperative to
preprocess the mediators by scaling them to have a mean of zero and a
variance of one prior to generating the outcome variables.</p>
</div>
<div id="parallel-computing-support" class="section level2">
<h2>Parallel Computing Support</h2>
<p>The <code>hima()</code> function supports <strong>parallel
computing</strong> to speed up high-dimensional mediation analysis,
especially when dealing with a large number of mediators.</p>
<div id="enabling-parallel-computing" class="section level3">
<h3>Enabling Parallel Computing</h3>
<p>To enable parallel computing, simply set <code>parallel = TRUE</code>
and specify the number of CPU cores to use via the <code>ncore</code>
argument:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">hima</span>(..., <span class="at">parallel =</span> <span class="cn">TRUE</span>, <span class="at">ncore =</span> <span class="dv">4</span>)</span></code></pre></div>
</div>
</div>
</div>
<div id="applications-and-examples" class="section level1">
<h1>Applications and Examples</h1>
<div id="load-the-hima-package" class="section level2">
<h2>Load the <code>HIMA</code> Package</h2>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">library</span>(HIMA)</span></code></pre></div>
</div>
<div id="continuous-outcome-analysis" class="section level2">
<h2>Continuous Outcome Analysis</h2>
<p>When analyzing continuous and normally distributed outcomes and
mediators, we can use the following code snippet:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">data</span>(ContinuousOutcome)</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>pheno_data <span class="ot">&lt;-</span> ContinuousOutcome<span class="sc">$</span>PhenoData</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>mediator_data <span class="ot">&lt;-</span> ContinuousOutcome<span class="sc">$</span>Mediator</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>hima_continuous.fit <span class="ot">&lt;-</span> <span class="fu">hima</span>(</span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>  Outcome <span class="sc">~</span> Treatment <span class="sc">+</span> Sex <span class="sc">+</span> Age,</span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>  <span class="at">data.pheno =</span> pheno_data,</span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a>  <span class="at">data.M =</span> mediator_data,</span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a>  <span class="at">mediator.type =</span> <span class="st">&quot;gaussian&quot;</span>,</span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a>  <span class="at">penalty =</span> <span class="st">&quot;DBlasso&quot;</span>,</span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a>  <span class="at">scale =</span> <span class="cn">FALSE</span> <span class="co"># Demo data is already standardized</span></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a>)</span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a><span class="fu">summary</span>(hima_continuous.fit, <span class="at">desc=</span><span class="cn">TRUE</span>) </span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a><span class="co"># `desc = TRUE` option to show the description of the output results</span></span></code></pre></div>
<p><code>penalty = &quot;DBlasso&quot;</code> is particularly effective at
identifying mediators with weaker signals compared to
<code>penalty = &quot;MCP&quot;</code>. However, using <code>DBlasso</code>
requires more computational time.</p>
</div>
<div id="efficient-hima" class="section level2">
<h2>Efficient HIMA</h2>
<p>For continuous and normally distributed mediators and outcomes, an
efficient HIMA method can be activated with the
<code>efficient = TRUE</code> option (penalty should be <code>MCP</code>
for the best results). This method may also provide greater statistical
power to detect mediators with weaker signals.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>hima_efficient.fit <span class="ot">&lt;-</span> <span class="fu">hima</span>(</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>  Outcome <span class="sc">~</span> Treatment <span class="sc">+</span> Sex <span class="sc">+</span> Age,</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>  <span class="at">data.pheno =</span> pheno_data,</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>  <span class="at">data.M =</span> mediator_data,</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>  <span class="at">mediator.type =</span> <span class="st">&quot;gaussian&quot;</span>,</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>  <span class="at">efficient =</span> <span class="cn">TRUE</span>,</span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>  <span class="at">penalty =</span> <span class="st">&quot;MCP&quot;</span>,</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>  <span class="at">scale =</span> <span class="cn">FALSE</span> <span class="co"># Demo data is already standardized</span></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a>)</span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a><span class="fu">summary</span>(hima_efficient.fit, <span class="at">desc=</span><span class="cn">TRUE</span>) </span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="co"># Note that the efficient HIMA is controlling FDR</span></span></code></pre></div>
<p>It is recommended to try different <code>penalty</code> options and
<code>efficient</code> option to find the best one for your data.</p>
</div>
<div id="binary-outcome-analysis" class="section level2">
<h2>Binary Outcome Analysis</h2>
<p>The package can handle binary outcomes based on logistic
regression:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">data</span>(BinaryOutcome)</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>pheno_data <span class="ot">&lt;-</span> BinaryOutcome<span class="sc">$</span>PhenoData</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>mediator_data <span class="ot">&lt;-</span> BinaryOutcome<span class="sc">$</span>Mediator</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>hima_binary.fit <span class="ot">&lt;-</span> <span class="fu">hima</span>(</span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>  Disease <span class="sc">~</span> Treatment <span class="sc">+</span> Sex <span class="sc">+</span> Age,</span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>  <span class="at">data.pheno =</span> pheno_data,</span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a>  <span class="at">data.M =</span> mediator_data,</span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a>  <span class="at">mediator.type =</span> <span class="st">&quot;gaussian&quot;</span>,</span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a>  <span class="at">penalty =</span> <span class="st">&quot;MCP&quot;</span>,</span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a>  <span class="at">scale =</span> <span class="cn">FALSE</span> <span class="co"># Demo data is already standardized</span></span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a>)</span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a><span class="fu">summary</span>(hima_binary.fit)</span></code></pre></div>
</div>
<div id="survival-outcome-analysis" class="section level2">
<h2>Survival Outcome Analysis</h2>
<p>For survival data, <code>HIMA</code> incorporates a Cox proportional
hazards approach. Here is an example of survival outcome analysis using
<code>HIMA</code>:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="fu">data</span>(SurvivalData)</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>pheno_data <span class="ot">&lt;-</span> SurvivalData<span class="sc">$</span>PhenoData</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>mediator_data <span class="ot">&lt;-</span> SurvivalData<span class="sc">$</span>Mediator</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>hima_survival.fit <span class="ot">&lt;-</span> <span class="fu">hima</span>(</span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a>  <span class="fu">Surv</span>(Time, Status) <span class="sc">~</span> Treatment <span class="sc">+</span> Sex <span class="sc">+</span> Age,</span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a>  <span class="at">data.pheno =</span> pheno_data,</span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a>  <span class="at">data.M =</span> mediator_data,</span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a>  <span class="at">mediator.type =</span> <span class="st">&quot;gaussian&quot;</span>,</span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a>  <span class="at">penalty =</span> <span class="st">&quot;DBlasso&quot;</span>,</span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a>  <span class="at">scale =</span> <span class="cn">FALSE</span> <span class="co"># Demo data is already standardized</span></span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a>)</span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a><span class="fu">summary</span>(hima_survival.fit)</span></code></pre></div>
</div>
<div id="microbiome-mediation-analysis" class="section level2">
<h2>Microbiome Mediation Analysis</h2>
<p>For compositional microbiome data, <code>HIMA</code> employs
isometric Log-Ratio transformations. Here is an example of microbiome
mediation analysis using <code>HIMA</code>:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="fu">data</span>(MicrobiomeData)</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>pheno_data <span class="ot">&lt;-</span> MicrobiomeData<span class="sc">$</span>PhenoData</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>mediator_data <span class="ot">&lt;-</span> MicrobiomeData<span class="sc">$</span>Mediator</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>hima_microbiome.fit <span class="ot">&lt;-</span> <span class="fu">hima</span>(</span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a>  Outcome <span class="sc">~</span> Treatment <span class="sc">+</span> Sex <span class="sc">+</span> Age,</span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a>  <span class="at">data.pheno =</span> pheno_data,</span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a>  <span class="at">data.M =</span> mediator_data,</span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a>  <span class="at">mediator.type =</span> <span class="st">&quot;compositional&quot;</span>,</span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a>  <span class="at">penalty =</span> <span class="st">&quot;DBlasso&quot;</span></span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a>)</span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a><span class="fu">summary</span>(hima_microbiome.fit)</span></code></pre></div>
</div>
<div id="quantile-mediation-analysis" class="section level2">
<h2>Quantile Mediation Analysis</h2>
<p>Perform quantile mediation analysis using the
<code>quantile = TRUE</code> option and specify <code>tau</code> for
desired quantile(s):</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="fu">data</span>(QuantileData)</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>pheno_data <span class="ot">&lt;-</span> QuantileData<span class="sc">$</span>PhenoData</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>mediator_data <span class="ot">&lt;-</span> QuantileData<span class="sc">$</span>Mediator</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>hima_quantile.fit <span class="ot">&lt;-</span> <span class="fu">hima</span>(</span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>  Outcome <span class="sc">~</span> Treatment <span class="sc">+</span> Sex <span class="sc">+</span> Age,</span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a>  <span class="at">data.pheno =</span> pheno_data,</span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a>  <span class="at">data.M =</span> mediator_data,</span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a>  <span class="at">mediator.type =</span> <span class="st">&quot;gaussian&quot;</span>,</span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a>  <span class="at">quantile =</span> <span class="cn">TRUE</span>,</span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a>  <span class="at">penalty =</span> <span class="st">&quot;MCP&quot;</span>,</span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a>  <span class="at">tau =</span> <span class="fu">c</span>(<span class="fl">0.3</span>, <span class="fl">0.5</span>, <span class="fl">0.7</span>),</span>
<span id="cb9-13"><a href="#cb9-13" tabindex="-1"></a>  <span class="at">scale =</span> <span class="cn">FALSE</span> <span class="co"># Demo data is already standardized</span></span>
<span id="cb9-14"><a href="#cb9-14" tabindex="-1"></a>)</span>
<span id="cb9-15"><a href="#cb9-15" tabindex="-1"></a><span class="fu">summary</span>(hima_quantile.fit)</span></code></pre></div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
